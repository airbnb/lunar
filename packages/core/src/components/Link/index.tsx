import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ButtonOrLink, { Props as ButtonOrLinkProps } from '../private/ButtonOrLink';
import Text from '../Text';

const sizingProp = mutuallyExclusiveTrueProps('small', 'large');
const stateProp = mutuallyExclusiveTrueProps('disabled', 'muted', 'inverted');

export type Props = ButtonOrLinkProps & {
  /** Display element as block. */
  block?: boolean;
  /** Display element as inline. */
  baseline?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Increase font size to large. */
  large?: boolean;
  /** Mark the link as muted. */
  muted?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
  /** Bold font. */
  bold?: boolean;
};

/** A standard link for... linking to things. */
export class Link extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    disabled: stateProp,
    inverted: stateProp,
    large: sizingProp,
    muted: stateProp,
    small: sizingProp,
  };

  static defaultProps = {
    baseline: false,
    block: false,
    bold: false,
    disabled: false,
    inverted: false,
    large: false,
    muted: false,
    small: false,
  };

  render() {
    const {
      cx,
      block,
      baseline,
      children,
      disabled,
      inverted,
      large,
      muted,
      small,
      bold,
      styles,
      ...restProps
    } = this.props;

    return (
      <Text inline={!block} baseline={baseline} small={small} large={large} bold={bold}>
        <ButtonOrLink
          {...restProps}
          disabled={disabled}
          className={cx(
            styles.link,
            inverted && styles.link_inverted,
            muted && styles.link_muted,
            disabled && styles.link_disabled,
            block && styles.link_block,
            baseline && styles.link_baseline,
          )}
        >
          {children}
        </ButtonOrLink>
      </Text>
    );
  }
}

export default withStyles(
  ({ color, pattern, transition }) => ({
    link: {
      ...pattern.resetButton,
      ...transition.box,
      color: color.core.primary[3],
      textAlign: 'left',
      verticalAlign: 'baseline',

      ':active': {
        outline: 'none',
      },

      ':hover': {
        color: color.core.primary[4],
        textDecoration: 'underline',
      },
    },

    link_block: {
      display: 'block',
      width: '100%',
    },

    link_baseline: {
      display: 'inline',
    },

    link_inverted: {
      color: color.accent.bg,

      ':hover': {
        color: color.accent.bgHover,
      },
    },

    link_muted: {
      color: color.core.neutral[3],

      ':hover': {
        color: color.core.neutral[4],
      },
    },

    link_disabled: {
      ...pattern.disabled,

      ':hover': {
        textDecoration: 'none',
      },
    },
  }),
  {
    extendable: true,
  },
)(Link);
