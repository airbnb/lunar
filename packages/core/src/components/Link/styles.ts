import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, pattern, transition }) => ({
  link: {
    ...pattern.resetButton,
    ...transition.box,
    color: color.core.primary[3],
    textAlign: 'left',
    verticalAlign: 'baseline',

    ':active': {
      outline: 'none',
    },

    ':hover': {
      color: color.core.primary[4],
      textDecoration: 'underline',
    },
  },

  link_block: {
    display: 'block',
    width: '100%',
  },

  link_baseline: {
    display: 'inline',
  },

  link_inverted: {
    color: color.accent.bg,

    ':hover': {
      color: color.accent.bgHover,
    },
  },

  link_muted: {
    color: color.core.neutral[3],

    ':hover': {
      color: color.core.neutral[4],
    },
  },

  link_disabled: {
    ...pattern.disabled,

    ':hover': {
      textDecoration: 'none',
    },
  },
});

export default styleSheet;

export { styleSheet };
