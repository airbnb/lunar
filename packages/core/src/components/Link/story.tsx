/* eslint-disable jsx-a11y/anchor-is-valid */

import React from 'react';
import { action } from '@storybook/addon-actions';
import IconAddAlt from '@airbnb/lunar-icons/lib/interface/IconAddAlt';
import Link from '.';

export default {
  title: 'Core/Link',
  parameters: {
    inspectComponents: [Link],
  },
};

export function aStandardLink() {
  return <Link href="https://github.com/airbnb/lunar">Link</Link>;
}

aStandardLink.story = {
  name: 'A standard link.',
};

export function aButtonLink() {
  return <Link onClick={action('onClick')}>Button</Link>;
}

aButtonLink.story = {
  name: 'A button link.',
};

export function withDifferentSizingSmallRegularDefaultAndLarge() {
  return (
    <>
      <Link small href="https://github.com/airbnb/lunar">
        Link
      </Link>{' '}
      <Link href="https://github.com/airbnb/lunar">Link</Link>{' '}
      <Link large href="https://github.com/airbnb/lunar">
        Link
      </Link>
    </>
  );
}

withDifferentSizingSmallRegularDefaultAndLarge.story = {
  name: 'With different sizing: small, regular (default), and large.',
};

export function withDifferentStatesMutedDisabledAndInverted() {
  return (
    <>
      <Link muted href="https://github.com/airbnb/lunar">
        Link
      </Link>{' '}
      <Link disabled href="https://github.com/airbnb/lunar">
        Link
      </Link>{' '}
      <Link inverted href="https://github.com/airbnb/lunar">
        Link
      </Link>
    </>
  );
}

withDifferentStatesMutedDisabledAndInverted.story = {
  name: 'With different states: muted, disabled, and inverted.',
};

export function withBeforeAndOrAfterIcons() {
  return (
    <>
      <div>
        <Link href="https://github.com/airbnb/lunar" beforeIcon={<IconAddAlt decorative />}>
          Link
        </Link>
        <br />
        <Link href="https://github.com/airbnb/lunar" afterIcon={<IconAddAlt decorative />}>
          Link
        </Link>
      </div>
    </>
  );
}

withBeforeAndOrAfterIcons.story = {
  name: 'With before and or after icons.',
};

export function stopClickAndDefaultEventsWhenDisabled() {
  return (
    <Link disabled href="https://github.com/airbnb/lunar" onClick={action('onClick')}>
      Link
    </Link>
  );
}

stopClickAndDefaultEventsWhenDisabled.story = {
  name: 'Stop click and default events when disabled.',
};

export function boldText() {
  return (
    <>
      <Link bold href="https://github.com/airbnb/lunar">
        Link
      </Link>
      <br />
      <Link bold onClick={action('onClick')}>
        Button
      </Link>
    </>
  );
}

boldText.story = {
  name: 'Bold text.',
};
