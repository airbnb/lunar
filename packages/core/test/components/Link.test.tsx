import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Link from '../../src/components/Link';

describe('<Link />', () => {
  it('errors when multiple states are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Link muted inverted href="#">
          Default
        </Link>,
      );
    }).toThrowError();
  });

  it('errors when multiple sizes are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Link small large href="#">
          Default
        </Link>,
      );
    }).toThrowError();
  });

  it('renders small', () => {
    const wrapper = shallowWithStyles(
      <Link small href="#">
        Small
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(
      <Link large href="#">
        Large
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const wrapper = shallowWithStyles(
      <Link disabled href="#">
        Disabled
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inverted', () => {
    const wrapper = shallowWithStyles(
      <Link inverted href="#">
        Inverted
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders muted', () => {
    const wrapper = shallowWithStyles(
      <Link muted href="#">
        Muted
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders the child component with an inline=false prop when block prop is passed', () => {
    const wrapper = shallowWithStyles(
      <Link block href="/foo">
        foo
      </Link>,
      true,
    );

    expect(wrapper.dive().prop('inline')).toBe(false);
  });

  it('renders bold', () => {
    const wrapper = shallowWithStyles(
      <Link bold href="#">
        Bold
      </Link>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
