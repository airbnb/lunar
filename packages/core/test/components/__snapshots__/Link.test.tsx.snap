// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders bold 1`] = `
<Text
  baseline={false}
  bold={true}
  inline={true}
  large={false}
  small={false}
>
  <ButtonOrLink
    className="link_vk1gl8"
    href="#"
  >
    Bold
  </ButtonOrLink>
</Text>
`;

exports[`<Link /> renders disabled 1`] = `
<Text
  baseline={false}
  bold={false}
  inline={true}
  large={false}
  small={false}
>
  <ButtonOrLink
    className="link_vk1gl8-o_O-link_disabled_q28bcx"
    disabled={true}
    href="#"
  >
    Disabled
  </ButtonOrLink>
</Text>
`;

exports[`<Link /> renders inverted 1`] = `
<Text
  baseline={false}
  bold={false}
  inline={true}
  large={false}
  small={false}
>
  <ButtonOrLink
    className="link_vk1gl8-o_O-link_inverted_1hj4kjb"
    href="#"
  >
    Inverted
  </ButtonOrLink>
</Text>
`;

exports[`<Link /> renders large 1`] = `
<Text
  baseline={false}
  bold={false}
  inline={true}
  large={true}
  small={false}
>
  <ButtonOrLink
    className="link_vk1gl8"
    href="#"
  >
    Large
  </ButtonOrLink>
</Text>
`;

exports[`<Link /> renders muted 1`] = `
<Text
  baseline={false}
  bold={false}
  inline={true}
  large={false}
  small={false}
>
  <ButtonOrLink
    className="link_vk1gl8-o_O-link_muted_74pra7"
    href="#"
  >
    Muted
  </ButtonOrLink>
</Text>
`;

exports[`<Link /> renders small 1`] = `
<Text
  baseline={false}
  bold={false}
  inline={true}
  large={false}
  small={true}
>
  <ButtonOrLink
    className="link_vk1gl8"
    href="#"
  >
    Small
  </ButtonOrLink>
</Text>
`;
